client/client.go

```diff
@@ -156,10 +156,120 @@ type tsoRequest struct {
 	dcLocation string
 }
 
+type tsoBatchController struct {
+	maxBatchSize         int
+	maxBatchWaitInterval time.Duration
+	// bestBatchSize is a dynamic size that changed based on the current batch effect.
+	bestBatchSize int
+
+	tsoRequestCh          chan *tsoRequest
+	collectedRequests     []*tsoRequest
+	collectedRequestCount int
+
+	batchStartTime time.Time
+}
+
+func newTSOBatchController(tsoRequestCh chan *tsoRequest, maxBatchSize int, maxBatchWaitInterval time.Duration) *tsoBatchController {
+	return &tsoBatchController{
+		maxBatchSize:          maxBatchSize,
+		maxBatchWaitInterval:  maxBatchWaitInterval,
+		bestBatchSize:         maxBatchSize,
+		tsoRequestCh:          tsoRequestCh,
+		collectedRequests:     make([]*tsoRequest, maxBatchSize+1),
+		collectedRequestCount: 0,
+	}
+}
+
+// fetchPendingRequests will start a new round of the batch collecting from the channel.
+// It returns true if everything goes well, otherwise false which means we should stop the service.
+func (tbc *tsoBatchController) fetchPendingRequests(ctx context.Context) error {
+	var firstTSORequest *tsoRequest
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case firstTSORequest = <-tbc.tsoRequestCh:
+	}
+	// Start to batch when the first TSO request arrives.
+	tbc.batchStartTime = time.Now()
+	tbc.collectedRequestCount = 0
+	tbc.pushRequest(firstTSORequest)
+
+	// This loop is for trying best to collect more requests, so we use `tbc.maxBatchSize` here.
+	for tbc.collectedRequestCount < tbc.maxBatchSize {
+		select {
+		case tsoReq := <-tbc.tsoRequestCh:
+			tbc.pushRequest(tsoReq)
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// break here can only break the select statement.
+			goto fetchMorePendingRequests
```

Comment:
use `break fetch` to break out of multi loop

Reply:
I think this is no different from `goto XXX`.

Reply:
When the code changes later, `goto` is more likely to cause problems that are not easy to find.